src/eduid_userdb/group_management/db.py

```diff
@@ -0,0 +1,174 @@
+# POSSIBILITY OF SUCH DAMAGE.
+#
+import datetime
+import logging
+from dataclasses import replace
+from operator import itemgetter
+from typing import Any, Dict, List, Optional
+
+from eduid_userdb.db import BaseDB
+from eduid_userdb.exceptions import DocumentOutOfSync
+from eduid_userdb.group_management.state import GroupInviteState
+
+logger = logging.getLogger(__name__)
+
+
+class GroupManagementInviteStateDB(BaseDB):
+    def __init__(self, db_uri: str, db_name: str = 'eduid_group_management', collection: str = 'group_invite_data'):
+        super(GroupManagementInviteStateDB, self).__init__(db_uri, db_name, collection=collection)
+        # Create an index so that invites for group_scim_id, email_address and role is unique
+        indexes = {
+            'unique-group-email-role': {
+                'key': [('group_scim_id', 1), ('email_address', 1), ('role', 1)],
+                'unique': True,
+            }
+        }
+        self.setup_indexes(indexes)
+
+    def get_state(
+        self, group_scim_id: str, email_address: str, role: str, raise_on_missing: bool = True
+    ) -> Optional[GroupInviteState]:
+        """
+        :param group_scim_id: Groups unique identifier
+        :param email_address: Invited email address
+        :param role: Group role
+        :param raise_on_missing: Raise exception if True else return None
+
+        :return: GroupInviteState instance
+        """
+        spec = {'group_scim_id': group_scim_id, 'email_address': email_address, 'role': role}
+        docs = list(self._get_documents_by_filter(spec, raise_on_missing=raise_on_missing))
+
+        if len(docs) > 1:
+            # Ex. multiple states for same group_scim_id, email address and roles matched
+            # This should not be possible but we have seen it happen
+            states = sorted(docs, key=itemgetter('modified_ts'))
+            state_to_keep = states.pop(-1)  # Keep latest state
+            for state in states:
+                self.remove_document(state['_id'])
+            return GroupInviteState.from_dict(state_to_keep)
+
+        return GroupInviteState.from_dict(docs[0])
+
+    def get_states_by_group_scim_id(self, group_scim_id: str, raise_on_missing: bool = True) -> List[GroupInviteState]:
+        """
+        Locate a state in the db given the state's group identifier.
+
+        :param group_scim_id: Groups unique identifier
+        :param raise_on_missing: Raise exception if True else return None
+
+        :return: List of GroupInviteState instances | None
+
+        :raise self.DocumentDoesNotExist: No document match the search criteria
+        """
+        ret: List[GroupInviteState] = list()
+        spec = {'group_scim_id': group_scim_id}
+        states = list(self._get_documents_by_filter(spec, raise_on_missing=raise_on_missing))
+
+        if len(states) == 0:
+            return ret
+
+        for state in states:
+            ret.append(GroupInviteState.from_dict(state))
+        return ret
```

Comment:
This function could be reduced to about 50% if you make instances using a list comprehension, but perhaps that's detrimental to readability:

return [GroupInviteState.from_dict(x) for x in states]